### ISSUE TITLE
Bundle product cache invalidation from simple product not working

### ISSUE THREAD

**Issue by paales** [magento/magento2#10236]
### Preconditions
1. 2.1.5

### Steps to reproduce

1. Create a bundle product with simple products
2. Save one of the simple product

### Expected result

1. http://cloud.h-o.nl/lYDp
2. This should be logged:
`"invalidateInfo":{"tags":["catalog_product_8469","catalog_product_8849"],"mode":"matchingTag"`

There 8469 is the simple product and 8849 is the 

### Actual result

1. The bundle product's cache isn't invalidated.
`"invalidateInfo":{"tags":["catalog_product_8469"],"mode":"matchingTag"`

### Workaround

```xml
<!-- plugin -->
    <type name="Magento\ConfigurableProduct\Model\Product\Cache\Tag\Configurable">
        <plugin sortOrder="1" name="hoKingdoConfigurable" type="Ho\Kingdo\Plugin\BundleProductCacheTagStrategy"/>
    </type>
```

```php
<?php
namespace Ho\Kingdo\Plugin;

use Magento\Catalog\Api\Data\ProductInterface;
use Magento\Framework\EntityManager\MetadataPool;

class BundleProductCacheTagStrategy
{

    /**
     * @var \Magento\Bundle\Model\ResourceModel\Selection
     */
    private $selection;

    /**
     * @var MetadataPool
     */
    private $metadataPool;

    public function __construct(
        \Magento\Bundle\Model\ResourceModel\Selection $selection,
        MetadataPool $metadataPool
    ) {
        $this->selection = $selection;
        $this->metadataPool = $metadataPool;
    }

    public function aroundGetTags(
        \Magento\ConfigurableProduct\Model\Product\Cache\Tag\Configurable $subject,
        \Closure $procede,
        $object
    ) {
        /** @var $object ProductInterface */
        if (! $object instanceof ProductInterface) {
            return $procede($object);
        }

        $result = $procede($object);

        foreach ($parentIds = $this->getParentIdsByChild($object->getId()) as $parentId) {
            $result[] = \Magento\Catalog\Model\Product::CACHE_TAG . '_' . $parentId;
        }

        return $result;
    }

    /**
     * Retrieve array of related bundle product ids by selection product id(s)
     *
     * @see \Magento\Bundle\Model\ResourceModel\Selection::getParentIdsByChild(), that method seems to be broken.
     * @param int|array $childId
     * @return array
     */
    public function getParentIdsByChild($childId)
    {
        $connection = $this->selection->getConnection();
        $metadata = $this->metadataPool->getMetadata(ProductInterface::class);
        $select = $connection->select()->distinct(
            true
        )->from(
            $this->selection->getMainTable(),
            ''
        )->join(
            ['e' => $this->metadataPool->getMetadata(ProductInterface::class)->getEntityTable()],
            'e.' . $metadata->getLinkField() . ' = ' .  $this->selection->getMainTable() . '.parent_product_id',
            ['e.entity_id as parent_product_id']
        )->where(
            $this->selection->getMainTable().'.product_id IN(?)',
            $childId
        );

        return $connection->fetchCol($select);
    }
}

```

**Comment by paales** [comment]
It seems that even with the above fix, there are still issues arising with cache-invalidation: When only the stock_item of a product is updated, it doesn't consider the product to be updated. This means that the cache wont be properly flushed when the stock of a product is updated.

To solve that issue for bundle products and configurable products, please take a look at: https://github.com/magento/magento2/issues/8009#issuecomment-315305029

**Comment by magento-engcom-team** [comment]
@paales, thank you for your report.
The issue is already fixed in 2.2.0
